pkg/option/config.go

```diff
@@ -1356,6 +1356,14 @@ func LogRegisteredOptions(vp *viper.Viper, entry *logrus.Entry) {
 	}
 }
 
+// FinalDaemonConfig is a alias for DaemonConfig used to differentiate what stage of daemon option init
+// the config is when it is provided.
+type FinalDaemonConfig DaemonConfig
```

Comment:
I am not sure I understand the intent for this. It's a parallel type and not an alias and it wraps an exported field. Is there an interface at play here?

Reply:
This seems more complicated than it needs to be. Could we instead just use the `DaemonConfig` type and have the provider for it also do the KPR etc. init/validation? (daemon/cmd/cells.go:91).

Reply:
@derailed The aliased type allows us to "re-provide" the same var (i.e. `*option.DaemonConfig`) but in a later state of initialization.  That is, we Provide DaemonConfig to hive for convenience but the underlying data is a global variable in pkg/option,  this var is modified at various points at both init/runtime for things like overriding required config for KPR.

This presents the problem that this breaks assumptions about Hive dependency ordering, as well as having configuration that cannot be finalized prior to runtime.

So by providing the FinalDaemonConfig _after_ the kpr init stuff, we provide a synchronization point to Hive where we know that the KPR init has already occured.

i.e.:
```go
// Arbitrary ordering, KPR init may not be complete prior to this: 
cell.Provide(func(_ *option.DaemonConfig) {})

// This depends on KPR init, so using fields mutated by that routine is "safe": 
cell.Provide(func(_ *option.FinalDaemonConfig) {})
```

Reply:
@joamaki this was the approach I tried initially, albeit with a "pre-init" DaemonConfig alias, but since this is global we might be able to just use the global var there and then provide DaemonConfig as a "post-kpr-init" state as the primary dependency.

Reply:
The "Final" part of this is actually misleading because this is only "final" for the KPR related fields so that might cause confusion - so I think maybe reworking this makes sense.

If these approaches prove unweidly, it may make sense to just step back and come up with a plan for how to fix this in its entirety (i.e. get rid of the global `*option.DaemonConfig`).  In the meantime we could also solve this by having the KPR init provider just provide some token dependency (ex. `postKPRInit{}`) where anything depending on that would be forced to wait for that to complete.